api/server/runner.go

```diff
@@ -57,7 +62,133 @@ var (
+
+	// GetCall can mod headers, assign an id, look up the route/app (cached),
+	// strip params, etc.
+	// this should happen ASAP to turn app name to app ID
+
+	// GetCall can mod headers, assign an id, look up the route/app (cached),
+	// strip params, etc.
+
+	call, err := s.agent.GetCall(
+		agent.WithWriter(&writer), // XXX (reed): order matters [for now]
+		agent.FromRequest(app, route, c.Request),
+	)
+	if err != nil {
+		return err
+	}
+	model := call.Model()
+	{ // scope this, to disallow ctx use outside of this scope. add id for handleV1ErrorResponse logger
+		ctx, _ := common.LoggerWithFields(c.Request.Context(), logrus.Fields{"id": model.ID})
+		c.Request = c.Request.WithContext(ctx)
+	}
+
+	if model.Type == "async" {
+		// TODO we should push this into GetCall somehow (CallOpt maybe) or maybe agent.Queue(Call) ?
+		if c.Request.ContentLength > 0 {
+			buf.Grow(int(c.Request.ContentLength))
+		}
+		_, err := buf.ReadFrom(c.Request.Body)
+		if err != nil {
+			return models.ErrInvalidPayload
+		}
+		model.Payload = buf.String()
+
+		err = s.enqueueDataAccess.Enqueue(c.Request.Context(), model)
+		if err != nil {
+			return err
+		}
+
+		c.JSON(http.StatusAccepted, map[string]string{"call_id": model.ID})
+		return nil
+	}
+
+	err = s.agent.Submit(call)
+	if err != nil {
+		// NOTE if they cancel the request then it will stop the call (kind of cool),
+		// we could filter that error out here too as right now it yells a little
+		if err == models.ErrCallTimeoutServerBusy || err == models.ErrCallTimeout {
+			// TODO maneuver
+			// add this, since it means that start may not have been called [and it's relevant]
+			c.Writer.Header().Add("XXX-FXLB-WAIT", time.Now().Sub(time.Time(model.CreatedAt)).String())
+		}
+		return err
+	}
+
+	// if they don't set a content-type - detect it
+	if writer.Header().Get("Content-Type") == "" {
+		// see http.DetectContentType, the go server is supposed to do this for us but doesn't appear to?
+		var contentType string
+		jsonPrefix := [1]byte{'{'} // stack allocated
+		if bytes.HasPrefix(buf.Bytes(), jsonPrefix[:]) {
+			// try to detect json, since DetectContentType isn't a hipster.
+			contentType = "application/json; charset=utf-8"
+		} else {
+			contentType = http.DetectContentType(buf.Bytes())
+		}
+		writer.Header().Set("Content-Type", contentType)
+	}
+
+	writer.Header().Set("Content-Length", strconv.Itoa(int(buf.Len())))
+
+	if writer.status > 0 {
+		c.Writer.WriteHeader(writer.status)
+	}
+	io.Copy(c.Writer, &writer)
+
+	return nil
+}
+
+// handleV1FunctionCall executes the function, for router handlers
+func (s *Server) handleHttpTriggerCall(c *gin.Context) {
+	err := s.handleTriggerHttpFunctionCall2(c)
+	if err != nil {
+		handleV1ErrorResponse(c, err)
+	}
+}
+
+// handleTriggerHttpFunctionCall2 executes the function and returns an error
+// Requires the following in the context:
+// * "app"
+// * "source"
+func (s *Server) handleTriggerHttpFunctionCall2(c *gin.Context) error {
```

Comment:
perhaps I missed it or it's still WIP (apologies), I don't see another handler that will allow invoking a function for any trigger type w/ cloud event data, only a handler for http triggers (which will mostly be serviced by an lb and can use the former, but we probably should have for convenience as well for 'getting started'). just didn't see anything and pointing this out in case, sure that something is in the works i'm just not sure of what :)

Reply:
yep, still WIP , haven't looked at cloud event endpoint - imagine that that would be a different URL